OSDOCS-19890 created zero-trust assembly/modules for multi-cluster (#112274)
openshift/openshift-docs: modules/zero-trust-manager-deploy-spire-multi-cluster.adoc

$ export OSSM_NS=istio-system
$ export OSSM_CNI=istio-cni

$ export CLUSTER_A_KUBECONFIG="/path/to/cluster-a/kubeconfig"
$ export CLUSTER_B_KUBECONFIG="/path/to/cluster-b/kubeconfig"

//Derived values
$ export CLUSTER_A_BASE_DOMAIN=$(oc get ingresses.config/cluster \
  -o jsonpath='{.spec.domain}' --kubeconfig "${CLUSTER_A_KUBECONFIG}")
$ export CLUSTER_B_BASE_DOMAIN=$(oc get ingresses.config/cluster \
  -o jsonpath='{.spec.domain}' --kubeconfig "${CLUSTER_B_KUBECONFIG}")

$ export CLUSTER_A_TRUST_DOMAIN="apps.${CLUSTER_A_BASE_DOMAIN##apps.}"
$ export CLUSTER_B_TRUST_DOMAIN="apps.${CLUSTER_B_BASE_DOMAIN##apps.}"

//If trust domains are the same as base domains (typical for OCP):
$ export CLUSTER_A_TRUST_DOMAIN="${CLUSTER_A_BASE_DOMAIN}"
$ export CLUSTER_B_TRUST_DOMAIN="${CLUSTER_B_BASE_DOMAIN}"

$ export CLUSTER_A=cluster-a
$ export CLUSTER_B=cluster-b
$ export NETWORK_A=network-a
$ export NETWORK_B=network-b

$ export FEDERATION_ENDPOINT_A="https://federation.${CLUSTER_A_BASE_DOMAIN}"
$ export FEDERATION_ENDPOINT_B="https://federation.${CLUSTER_B_BASE_DOMAIN}"

$ echo "Cluster A: ${CLUSTER_A_BASE_DOMAIN} (trust: ${CLUSTER_A_TRUST_DOMAIN})"

> $ echo "Cluster B: ${CLUSTER_B_BASE_DOMAIN} (trust: ${CLUSTER_B_TRUST_DOMAIN})"

Reviewer: **CRITICAL: SpireServer CR has non-existent `trustDomain` and `clusterName` fields**

`trustDomain` and `clusterName` do NOT exist on `SpireServerSpec`. These fields only exist on `ZeroTrustWorkloadIdentityManagerSpec`. Kubernetes will silently prune them — they have no effect.

**SpireServerSpec actual fields**: `logLevel`, `logFormat`, `jwtIssuer` (Required), `caValidity`, `defaultX509Validity`, `defaultJWTValidity`, `caKeyType`, `jwtKeyType`, `keyManager`, `caSubject` (Required), `persistence` (Required), `datastore` (Required), `federation`, `upstreamAuthority`

The operator controllers read `trustDomain` and `clusterName` from the parent ZTWIM CR at reconcile time:
```go
// pkg/controller/spire-server/configmap.go
"trust_domain": ztwim.Spec.TrustDomain,
```

**Fix**: Remove `trustDomain` and `clusterName` from the SpireServer CR example. They belong only on the `ZeroTrustWorkloadIdentityManager` CR (which correctly has them above).

Ref: [SpireServerSpec types](https://github.com/openshift/zero-trust-workload-identity-manager/blob/main/api/v1alpha1/spire_server_config_types.go#L30) | [ZTWIMSpec types](https://github.com/openshift/zero-trust-workload-identity-manager/blob/main/api/v1alpha1/zero_trust_workload_identity_manager_types.go#L114)

Author: Done based on the ref provided above

----

. Deploy SPIRE with federation enabled:

.. Deploy the SPIRE operands on cluster A:
+
[source,terminal]
----
$ export KUBECONFIG="${CLUSTER_A_KUBECONFIG}"
export JWT_ISSUER_A="https://oidc-discovery.${CLUSTER_A_BASE_DOMAIN}"

oc apply -f - <<EOF
apiVersion: operator.openshift.io/v1alpha1
kind: ZeroTrustWorkloadIdentityManager
metadata:
  name: cluster
spec:
  trustDomain: ${CLUSTER_A_TRUST_DOMAIN}
  clusterName: ${CLUSTER_A}
---
apiVersion: operator.openshift.io/v1alpha1
kind: SpireServer
metadata:
  name: cluster
spec:
  trustDomain: ${CLUSTER_A_TRUST_DOMAIN}
  clusterName: ${CLUSTER_A}
  caSubject:

>     commonName: ${CLUSTER_A_TRUST_DOMAIN}

Reviewer: **CRITICAL: SpireAgent CR has non-existent `trustDomain` and `clusterName` fields**

Same issue as the SpireServer CR above. `SpireAgentSpec` does NOT have `trustDomain` or `clusterName`.

**SpireAgentSpec actual fields**: `socketPath` (default: `/run/spire/agent-sockets`), `logLevel`, `logFormat`, `nodeAttestor`, `workloadAttestors`

**Fix**: Remove `trustDomain` and `clusterName` from the SpireAgent CR.

Ref: [SpireAgentSpec types](https://github.com/openshift/zero-trust-workload-identity-manager/blob/main/api/v1alpha1/spire_agent_config_types.go#L26)

    country: "US"
    organization: "RH"
  persistence:
    type: pvc
    size: "2Gi"
    accessMode: ReadWriteOncePod
  datastore:
    databaseType: sqlite3
    connectionString: "/run/spire/data/datastore.sqlite3"
    maxOpenConns: 100
    maxIdleConns: 2
    connMaxLifetime: 3600
  jwtIssuer: ${JWT_ISSUER_A}
  federation:

>     enabled: true

Reviewer 2: this field also doesn't exist

Reviewer 3: does this comment still hold true?

    bundleEndpoint:
      address: "0.0.0.0"
      port: 8443

> ---

Reviewer 2: incorrect fields inside bundleEndpoint

> apiVersion: operator.openshift.io/v1alpha1

Reviewer: **CRITICAL: SpireOIDCDiscoveryProvider has non-existent `trustDomain` field**

`trustDomain` does NOT exist on `SpireOIDCDiscoveryProviderSpec`. Only `jwtIssuer` is a valid required field here.

**SpireOIDCDiscoveryProviderSpec actual fields**: `logLevel`, `logFormat`, `csiDriverName`, `jwtIssuer` (Required), `replicaCount`, `managedRoute`, `externalSecretRef`

**Fix**: Remove `trustDomain` from the SpireOIDCDiscoveryProvider CR.

Ref: [SpireOIDCDiscoveryProviderSpec types](https://github.com/openshift/zero-trust-workload-identity-manager/blob/main/api/v1alpha1/spire_oidc_discovery_provider_types.go#L26)

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-spire-multi-cluster_{context}"]
= Deploying {spire-full} with federation for multi-cluster integration

[role="_abstract"]
Deploy {spire-full} with federation capabilities on each cluster to enable cross-cluster trust bundle exchange and workload identity validation across multiple {product-title} clusters.

.Prerequisites

* You have two {product-title} clusters (4.x) with network connectivity between them.
* You have installed {zero-trust-full} on both clusters.
* The `oc` CLI is configured with access to both clusters.
* You have installed `istioctl`.

> * You have `helm` installed. This is used for gateway deployment.

Reviewer 3: ```suggestion
* You have installed `helm`. This is used for gateway deployment.
```

* You have Istio version >= 1.29.2

.Procedure

. Define the following environment variables:
+
[source,terminal]
----
$ export ZTWIM_NS=zero-trust-workload-identity-manager
$ export OSSM_NS=istio-system
$ export OSSM_CNI=istio-cni

$ export CLUSTER_A_KUBECONFIG="/path/to/cluster-a/kubeconfig"
$ export CLUSTER_B_KUBECONFIG="/path/to/cluster-b/kubeconfig"

$ export CLUSTER_A_BASE_DOMAIN=$(oc get ingresses.config/cluster \
  -o jsonpath='{.spec.domain}' --kubeconfig "${CLUSTER_A_KUBECONFIG}")
$ export CLUSTER_B_BASE_DOMAIN=$(oc get ingresses.config/cluster \
  -o jsonpath='{.spec.domain}' --kubeconfig "${CLUSTER_B_KUBECONFIG}")

$ export CLUSTER_A_TRUST_DOMAIN="${CLUSTER_A_BASE_DOMAIN}"
$ export CLUSTER_B_TRUST_DOMAIN="${CLUSTER_B_BASE_DOMAIN}"

$ export CLUSTER_A=cluster-a
$ export CLUSTER_B=cluster-b
$ export NETWORK_A=network-a
$ export NETWORK_B=network-b

$ export FEDERATION_ENDPOINT_A="https://federation.${CLUSTER_A_BASE_DOMAIN}"
$ export FEDERATION_ENDPOINT_B="https://federation.${CLUSTER_B_BASE_DOMAIN}"

$ export JWT_ISSUER_A="https://oidc-discovery.${CLUSTER_A_BASE_DOMAIN}"

> $ export JWT_ISSUER_B="https://oidc-discovery.${CLUSTER_B_BASE_DOMAIN}"

Reviewer 3: Group these commands and have individual commands setting that group of variables in every step.

----

. Deploy SPIRE with federation enabled:

> .. Create the `ZeroTrustWorkloadIdentityManager` CR on Cluster A:

Reviewer 3: by running the following command

+
[source,terminal]
----
$ export KUBECONFIG="${CLUSTER_A_KUBECONFIG}"

$ oc apply -f - <<EOF
apiVersion: operator.openshift.io/v1alpha1
kind: ZeroTrustWorkloadIdentityManager
metadata:
  name: cluster
spec:
  trustDomain: ${CLUSTER_A_TRUST_DOMAIN}
  clusterName: ${CLUSTER_A}
  bundleConfigMap: "spire-bundle"
EOF
----

> .. Create the `SpireServer` CR:

Reviewer 3: by running the following command

+

> [source,yaml]

Reviewer 3: ```suggestion
[source,terminal]
```

----
$ cat <<EOF | oc apply -f -
apiVersion: operator.openshift.io/v1alpha1
kind: SpireServer
metadata:
  name: cluster
spec:
  logLevel: "info"
  logFormat: "text"
  jwtIssuer: $JWT_ISSUER_A
  caValidity: "24h"
  defaultX509Validity: "1h"
  defaultJWTValidity: "5m"
  caKeytype: "rsa-2048"
  jwtKeyType: "rsa-2048"
  keyManager: ""
  caSubject:
    country: "US"
    organization: "RH"
    commonName: "SPIRE Server CA"
  persistence:
    size: "5Gi"
    accessMode: "ReadWriteOnce"
  datastore:
    databaseType: "sqlite3"
    connectionString: "/run/spire/data/datastore.sqlite3"
    tlsSecretName: ""
    maxOpenConns: 100
    maxIdleConns: 10
    connMaxLifetime: 0
    disableMigration: "false"
  federation:
    enabled: true
    bundleEndpoint:
      address: "0.0.0.0"
      port: 8443
EOF
----

> .. Create the `SpireAgent` CR:

Reviewer 3: by running the following command

+

> [source,yaml]

Reviewer 3: ```suggestion
[source,terminal]
```

----
$ cat <<EOF | oc apply -f -
apiVersion: operator.openshift.io/v1alpha1
kind: SpireAgent
metadata:
  name: cluster
spec:
  socketPath: "/run/spire/agent-sockets"
  logLevel: "info"
  logFormat: "text"
  nodeAttestor:
    k8sPSATEnabled: "true"
  workloadAttestors:
    k8sEnabled: "true"
    workloadAttestorsVerification:
      type: "auto"
      hostCertBasePath: "/etc/kubernetes"
      hostCertFileName: "kubelet-ca.crt"
    useNewContainerLocator: "true"
EOF
----

> .. Create the `SpiffeCSIDriver` CR:

Reviewer 3: by running the following command

+

> [source,yaml]

Reviewer 3: ```suggestion
[source,terminal]
```

----
$ cat <<EOF | oc apply -f -
apiVersion: operator.openshift.io/v1alpha1
kind: SpiffeCSIDriver
metadata:
  name: cluster
spec:
  agentSocketPath: "/run/spire/agent-sockets"
  pluginName: csi.spiffe.io
EOF
----

> .. Create the `SpireOIDCDiscoveryProvider` CR:

Reviewer 3: by running the following command

+

> [source,yaml]

Reviewer 3: ```suggestion
[source,terminal]
```

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-spire-multi-cluster_{context}"]
= Deploying {spire-full} with federation for multi-cluster integration

[role="_abstract"]
Deploy {spire-full} with federation capabilities on each cluster to enable cross-cluster trust bundle exchange and workload identity validation across multiple {product-title} clusters.

.Prerequisites

* You have two {product-title} clusters (4.x) with network connectivity between them.
* You have installed {zero-trust-full} on both clusters.

> * The `oc` CLI is configured with access to both clusters.

Reviewer 4: ```suggestion
* The {oc-first} is configured with access to both clusters.
```

> * You have installed `istioctl`.

Reviewer 4: ```suggestion
* You have installed the `istioctl` CLI tool.
```

> * You have installed `helm`. This is used for gateway deployment.

Reviewer 4: is this a CLI tool, or do you mean Helm?

> * You have Istio version >= 1.29.2

Reviewer 4: ```suggestion
* You have Istio version 1.29.2 or later.
```
is the proper form

.Procedure

. Export the namespace variables by running the following commands:
+
[source,terminal]
----
$ export ZTWIM_NS=zero-trust-workload-identity-manager
----
+
[source,terminal]
----
$ export OSSM_NS=istio-system
----
+
[source,terminal]
----
$ export OSSM_CNI=istio-cni
----

. Export the kubeconfig file paths for Cluster A and Cluster B by running the following commands:
+
[source,terminal]
----
$ export CLUSTER_A_KUBECONFIG="/path/to/cluster-a/kubeconfig"
----
+
[source,terminal]
----
$ export CLUSTER_B_KUBECONFIG="/path/to/cluster-b/kubeconfig"
----

. Set the base domain environment variables for Cluster A and Cluster B by running the following commands:
+
[source,terminal]
----
$ export CLUSTER_A_BASE_DOMAIN=$(oc get ingresses.config/cluster \
  -o jsonpath='{.spec.domain}' --kubeconfig "${CLUSTER_A_KUBECONFIG}")
----
+
[source,terminal]
----
$ export CLUSTER_B_BASE_DOMAIN=$(oc get ingresses.config/cluster \
  -o jsonpath='{.spec.domain}' --kubeconfig "${CLUSTER_B_KUBECONFIG}")
----

> . Export the trust domain environment variables from each cluster's base domain by running the following commands:

Reviewer 4: ```suggestion
. Export the trust domain environment variables from each cluster base domain by running the following commands:
```

Reviewer 5: Agree with Shauna's suggestion. Typically, we do not use contractions in the docs.

https://redhat-documentation.github.io/supplementary-style-guide/#contractions

+
[source,terminal]
----
$ export CLUSTER_A_TRUST_DOMAIN="${CLUSTER_A_BASE_DOMAIN}"
----

> [source,terminal]

Reviewer 4: ```suggestion
+
[source,terminal]
```
this one is misaligned in the preview

----
$ export CLUSTER_B_TRUST_DOMAIN="${CLUSTER_B_BASE_DOMAIN}"
----

> . Define the cluster and network environment variables by running the following commands:

Reviewer 4: and that caused this step to be renumbered 1; so just double check after you fix the previous command block

+
[source,terminal]
----
$ export CLUSTER_A=cluster-a
----
+
[source,terminal]
----
$ export CLUSTER_B=cluster-b
----
+
[source,terminal]
----
$ export NETWORK_A=network-a
----
+
[source,terminal]
----
$ export NETWORK_B=network-b
----

. Export the federation endpoint URLs for Cluster A and Cluster B by running the following commands:
+
[source,terminal]
----
$ export FEDERATION_ENDPOINT_A="https://federation.${CLUSTER_A_BASE_DOMAIN}"
----
+
[source,terminal]
----
$ export FEDERATION_ENDPOINT_B="https://federation.${CLUSTER_B_BASE_DOMAIN}"
----

> . Set the JWT issuer environment variables for Cluster A and Cluster B by running the following commands:

Reviewer 4: ```suggestion
. Set the JWT issuer environment variables for Cluster A by running the following command:
```

+
[source,terminal]
----
$ export JWT_ISSUER_A="https://oidc-discovery.${CLUSTER_A_BASE_DOMAIN}"
----

> +

Reviewer 4: ```suggestion

..Set the JWT issuer environment variables for Cluster B by running the following command:
```
see Shubha's comment

[source,terminal]
----
$ export JWT_ISSUER_B="https://oidc-discovery.${CLUSTER_B_BASE_DOMAIN}"
----

> . Deploy SPIRE with federation enabled:

Reviewer 4: procedure is too long; break this into its own module here
